api/client/proxy.go

```diff
@@ -98,6 +122,40 @@ func DialProxyWithDialer(ctx context.Context, proxyURL *url.URL, addr string, di
 	}, nil
 }
 
+// dialProxyWithSOCKSDialer creates a connection to a server via a SOCKS Proxy.
+func dialProxyWithSOCKSDialer(
+	ctx context.Context,
+	proxyURL *url.URL,
+	addr string,
+	dialer *net.Dialer,
+) (net.Conn, error) {
+	if proxyURL == nil {
+		return nil, trace.BadParameter("missing proxy url")
+	}
+
+	var proxyAuth *proxy.Auth
+	if proxyURL.User != nil {
+		password, _ := proxyURL.User.Password()
+		proxyAuth = &proxy.Auth{
+			User:     proxyURL.User.Username(),
+			Password: password,
+		}
+	}
+
+	socksDialer, err := proxy.SOCKS5("tcp", proxyURL.Host, proxyAuth, dialer)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	conn, err := socksDialer.Dial("tcp", addr)
+	if err != nil {
+		log.Warnf("unable to proxy connection: %v", err)
+		return nil, trace.ConvertSystemError(err)
```

Comment:
nit: error should be either logger or returned. Not both. Otherwise, the caller will probably do the same, and we will have duplicated logs.